ftlib/consensus/shared_storage/impl.py

```diff
@@ -0,0 +1,206 @@
+
+    def get_rank_size(self, maddr=False):
+        ipc_dict = {
+            ip: counter for ip, counter in zip(self._ips, self._counts)
+        }
+        max_count = max(ipc_dict.values())
+        assert max_count == self._count
+        ips = [ip for ip, count in ipc_dict.items() if count == max_count]
+        rank, size, master_addr = rank_assign_scheme(ips=ips, my_ip=ip_address)
+
+        if rank == 0:
+            try:
+                self._server.start()
+            except Exception as e:
+                logging.warning(str(e))
+            else:
+                logging.info("server starts as root")
+        else:
+            try:
+                self._server.stop(2)
+            except Exception as e:
+                logging.warning(str(e))
+            else:
+                logging.info("server stops")
+
+        if maddr:
+            return rank, size, master_addr
+        else:
+            return rank, size
+
+    def average_failure(self):
+        pass
+
+    def average_success(self):
+        self._count = self._count + 1
+
+    def new_member_join(self, status=True):
+        self._ftlib.lock()
+        logging.info("new member join!")
+        self._ftlib._new_member_join = status
+        self._ftlib._skip_allreduce = False
+        try:
+            self._server.stop(1)
+        except Exception as e:
+            logging.info(str(e))
+        else:
+            logging.info("gRPC server stops when new member joins")
+        self._ftlib.unlock()
+
+    def get_count(self):
+        return self._count
+
+    def _become_root(self):
+        if self._rank == 0:
+            return
+        self._rank = 0
+        try:
+            self._server.start()
+        except ValueError as e:
+            logging.error(str(e))
+        else:
+            logging.info("root rank started gRPC server")
+
+    def _report_join(self, ips):
+        for ip in ips:
+            channel = grpc.insecure_channel(
+                "{ip}:{port}".format(ip=ip, port=self._port)
+            )
+            try:
+                grpc.channel_ready_future(channel).result(timeout=2)
+            except Exception as e:
+                logging.warning(str(e))
+                return False
+            else:
+                stub = communicate_pb2_grpc.ReportStub(channel)
+            try:
+                response = stub.GetGroupStatus(
+                    communicate_pb2.GroupStatusRequest(), timeout=2
+                )
+            except Exception as e:
+                logging.warning(str(e))
+            else:
+                logging.info("counter retreived")
+                self._count = response.counter
+                self._io_tool.register_ip(ip_address, self._count)
+                return True
+        logging.warning("counter not retreived")
+        self._io_tool.register_ip(ip_address, self._count)
+        return False
+
+
+atexit.register(clean_my_ip_file, shared_path, ip_address)
```

Comment:
If this file gets sourced/loaded, this line will get executed. Is this expected?

Reply:
Yes, I believe `clean_my_ip_file` will be registered every time the file loaded. It's not expected. Any suggestion on how I can improve it?

Reply:
I am not sure if I understand the purpose of this. Why are you using `atexit` here instead of explicitly calling some kind of method to register when they are used? Maybe moving the variables `hostname` and `ip_address` to `SharedStorage` as member variables and then try to use `atexit.register` as a decorator ([example](https://docs.python.org/3/library/atexit.html#atexit-example)) instead?

Reply:
I see. Let me try it and update the script.

Reply:
PTAL